src/libmeasurement_kit/ffi.cpp

```diff
@@ -0,0 +1,99 @@
+// Part of Measurement Kit <https://measurement-kit.github.io/>.
+// Measurement Kit is free software under the BSD license. See AUTHORS
+// and LICENSE for more information on the copying conditions.
+
+#define MK_ENGINE_INTERNALS // we need to access engine internals
+
+#include <measurement_kit/ffi.h>
+
+#include <limits.h>
+#include <string.h>
+
+#include <exception>
+#include <new>
+#include <string>
+
+#include <measurement_kit/common/nlohmann/json.hpp>
+#include <measurement_kit/engine.h>
+
+struct mk_event_ : public std::string {
+    using std::string::string;
+};
+
+static mk_event_t *mk_event_create(const nlohmann::json &json) noexcept {
+    return new (std::nothrow) mk_event_t{json.dump().data()};
+}
+
+const char *mk_event_serialize(mk_event_t *event) noexcept {
+    return (event) ? event->data() : nullptr;
+}
+
+void mk_event_destroy(mk_event_t *event) noexcept {
+    delete event; // handles nullptr
+}
+
+struct mk_task_ : mk::engine::Task {
+    using mk::engine::Task::Task;
+};
+
+mk_task_t *mk_task_start(const char *settings) noexcept {
+    return mk_task_start_ex(settings, nullptr, -1);
+}
+
+static void safe_copy(const char *reason, char *errbuf, size_t errbuf_siz) {
+    if (errbuf != nullptr && errbuf_siz > 0) {
+        size_t size = strlen(reason);
+        if (size > SIZE_MAX - 1) {
+            return;
+        }
+        size += 1; // account for the final '\0'
```

Comment:
Maybe it's cleaner (and more readable) to increase the size and then do the `if` statement. This way you can avoid adding and removing the - 1.

Though perhaps you are doing that to avoid int overflow in `size`.

Reply:
> Though perhaps you are doing that to avoid int overflow in size.

Yes, indeed!